src/Umbraco.Core/Persistence/Repositories/ContentRepository.cs

```diff
@@ -1126,31 +1126,10 @@ private string EnsureUniqueNodeName(int parentId, string nodeName, int id = 0)
             if (EnsureUniqueNaming == false)
                 return nodeName;
 
-            var sql = new Sql();
-            sql.Select("*")
-               .From<NodeDto>()
-               .Where<NodeDto>(x => x.NodeObjectType == NodeObjectTypeId && x.ParentId == parentId && x.Text.StartsWith(nodeName));
-
-            int uniqueNumber = 1;
-            var currentName = nodeName;
-
-            var dtos = Database.Fetch<NodeDto>(sql);
-            if (dtos.Any())
-            {
-                var results = dtos.OrderBy(x => x.Text, new SimilarNodeNameComparer());
-                foreach (var dto in results)
-                {
-                    if (id != 0 && id == dto.NodeId) continue;
-
-                    if (dto.Text.ToLowerInvariant().Equals(currentName.ToLowerInvariant()))
-                    {
-                        currentName = nodeName + string.Format(" ({0})", uniqueNumber);
-                        uniqueNumber++;
-                    }
-                }
-            }
+            var names = Database.Fetch<SimilarNodeName>("SELECT id, text AS name FROM umbracoNode WHERE nodeObjectType=@objectType AND parentId=@parentId",

```

Comment:
Interestingly, here we are `Fetch` ing which means loading the entire data set into memory before continuing, however the call below to SimilarNodeName.GetUniqueName takes in an IEnumerable and it would seem that this enumerable is not double iterated. So... we can change this to `Query` instead of `Fetch` which will create a forward reading db cursor and load less into memory. 

This is 'in theory',  inside of SimilarNodeName.GetUniqueName there's a custom sorter but that will only compare the current and the next so i 'think' the read forward db cursor may still work in this scenario. I'll test.